common/context/src/main/java/io/helidon/common/context/ContextAwareExecutorImpl.java

```diff
@@ -0,0 +1,134 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+class ContextAwareExecutorImpl implements ContextAwareExecutorService {
+    private final ExecutorService delegate;
+
+    ContextAwareExecutorImpl(ExecutorService toWrap) {
+        this.delegate = toWrap;
+    }
+
+    @Override
+    public void shutdown() {
+        delegate.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        return delegate.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return delegate.isShutdown();
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return delegate.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return delegate.awaitTermination(timeout, unit);
+    }
+
+    @Override
+    public <T> Future<T> submit(Callable<T> task) {
+        return delegate.submit(wrap(task));
+    }
+
+    @Override
+    public <T> Future<T> submit(Runnable task, T result) {
+        return delegate.submit(wrap(task), result);
+    }
+
+    @Override
+    public Future<?> submit(Runnable task) {
+        return delegate.submit(wrap(task));
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+        return delegate.invokeAll(wrap(tasks));
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+            throws InterruptedException {
+        return delegate.invokeAll(wrap(tasks), timeout, unit);
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
+        return delegate.invokeAny(wrap(tasks));
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        return delegate.invokeAny(wrap(tasks), timeout, unit);
+    }
+
+    @Override
+    public void execute(Runnable command) {
+        delegate.execute(wrap(command));
+    }
+
+    @Override
+    public ExecutorService unwrap() {
+        return delegate;
+    }
+
+    protected <T> Collection<? extends Callable<T>> wrap(Collection<? extends Callable<T>> tasks) {
+        return tasks.stream()
+                .map(this::wrap)
+                .collect(Collectors.toList());
+
+    }
+
+    protected  <T> Callable<T> wrap(Callable<T> task) {
+        Optional<Context> context = Contexts.context();
+
+        if (context.isPresent()) {
+            return () -> Contexts.inContext(context.get(), task);
+        } else {
+            return task;
+        }
+    }
+
+    protected Runnable wrap(Runnable command) {
```

Comment:
Not so much a comment, but more of a general question on style preference.

This whole method *could* be replaced with a "functional" one-liner:
```
return Contexts.context()
                .map(ctx -> () -> Contexts.inContext(context.get(), command))
                .orElse(command);
```
Do we have a preference for one style over another?

The functional approach is slightly shorter, but the code below, as currently written, is easier to debug.

Reply:
We have had a few discussions about this in the team, some feel the use of `map` in this context is somewhat artificial and not necessarily easier to read. Personally, I prefer the use of _actual_ language statements (which tends to reduce the number of capturing lambdas as well), but in the case above would have used a conditional expression `? :`.

Reply:
There is an issue we can add this to - #648 
In this specific case the lambda is not nice, as it does `ctx -> () -> ...` - this double lambding is not easy to read (e.g. a lambda that returns a lambda), and in this case the current code is nicer to read - there was actually a review comment asking me to switch from lambda to if/then/else